scripts/util/HelperNode.lua

```diff
@@ -0,0 +1,110 @@
+--[[
+This file is part of Courseplay (https://github.com/Courseplay/FS22_Courseplay)
+Copyright (C) 2024 Courseplay Dev Team
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+]]--
+
+-- A helper node is a temporary node to help with relative vehicle position calculations.
+-- It wraps the creation of the Giant's engine node.
+-- NOTE: it must be destroyed explicitly to avoid node leak.
+
+---@class HelperNode
+HelperNode = CpObject()
+
+--- Create a new helper node, linking it to the given rootNode.
+---@param name string
+---@param rootNode number|nil
+function HelperNode:init(name, rootNode)
+    self.node = createTransformGroup(name)
+    self.rootNode = rootNode
+    link(self.rootNode, self.node)
+end
+
+--- Safely destroy a node
+function HelperNode:destroy()
+    if self.node and entityExists(self.node) then
+        unlink(self.node)
+        delete(self.node)
+    end
+end
+
+--- Place the node at the given position and rotation.
+---@param x number
+---@param y number if the node is linked to the terrain, this is relative to the terrain height at x, z
+---@param z number
+---@param yRotation number|nil Rotation set only if not nil
+function HelperNode:place(x, y, z, yRotation)
+    setTranslation(self.node, x, y, z)
+    if yRotation then
+        setRotation(self.node, 0, yRotation, 0)
+    end
+end
+
+---@param text string|nil
+function HelperNode:draw(text)
+    if entityExists(self.node) then
+        DebugUtil.drawDebugNode(self.node, text or getName(self.node), false, 0)
+    end
+end
+
+-- A helper node that is linked to the terrain root node, and the y coordinate is always
+-- the terrain height at the x, z position.
+---@class HelperTerrainNode : HelperNode
+HelperTerrainNode = CpObject(HelperNode)
+
+--- Create a new terrain helper node, linking it to the terrain root node.
+---@param name string
+function HelperTerrainNode:init(name)
+    HelperNode.init(self, name, g_currentMission.terrainRootNode)
+end
+
+--- Place the node at the given position and rotation.
+---@param x number
+---@param y number if the node is linked to the terrain, this is relative to the terrain height at x, z
+---@param z number
+---@param yRotation number|nil Rotation set only if not nil
+function HelperTerrainNode:place(x, y, z, yRotation)
```

Comment:
I would prefer this functionality in a separate function like so maybe:
```lua
---@param yRotation number|nil
---@param yOffset number|nil
function HelperNode:placeTerrain(x, z, yRotation, yOffset)

end
```

Reply:
Was struggling with this as well, probably we don't even need a generic HelperNode (at least as of now, I'm not aware of a use case), and should only implement a HelperTerrainNode.